platform/posix/transport/src/mbedtls_pkcs11_posix.c

```diff
@@ -0,0 +1,958 @@
+    }
+
+    if( returnStatus == MBEDTLS_PKCS11_SUCCESS )
+    {
+        returnStatus = configureMbedtlsSniAlpn( pMbedtlsPkcs11Context, pMbedtlsPkcs11Credentials, pHostName );
+    }
+
+    if( returnStatus == MBEDTLS_PKCS11_SUCCESS )
+    {
+        /* Initialize the MbedTLS secured connection context. */
+        mbedtlsError = mbedtls_ssl_setup( &( pMbedtlsPkcs11Context->context ),
+                                          &( pMbedtlsPkcs11Context->config ) );
+
+        if( mbedtlsError != 0 )
+        {
+            LogError( ( "Failed to set up MbedTLS SSL context: mbedTLSError= %s : %s.",
+                        mbedtlsHighLevelCodeOrDefault( mbedtlsError ),
+                        mbedtlsLowLevelCodeOrDefault( mbedtlsError ) ) );
+            returnStatus = MBEDTLS_PKCS11_INTERNAL_ERROR;
+        }
+    }
+
+    if( returnStatus == MBEDTLS_PKCS11_SUCCESS )
+    {
+        /* Set the underlying IO for the TLS connection. */
+        mbedtls_ssl_set_bio( &( pMbedtlsPkcs11Context->context ),
+                             ( void * ) &( pMbedtlsPkcs11Context->socketContext ),
+                             mbedtls_net_send,
+                             mbedtls_net_recv,
+                             mbedtls_net_recv_timeout );
+
+        returnStatus = configureMbedtlsFragmentLength( pMbedtlsPkcs11Context );
+    }
+
+    if( returnStatus != MBEDTLS_PKCS11_SUCCESS )
+    {
+        contextFree( pMbedtlsPkcs11Context );
+    }
+    else
+    {
+        LogInfo( ( "(Network connection %p) TLS handshake successful.", pNetworkContext ) );
+    }
+
+    return returnStatus;
+}
+
+/*-----------------------------------------------------------*/
+
+static MbedtlsPkcs11Status_t configureMbedtlsCertificates( MbedtlsPkcs11Context_t * pMbedtlsPkcs11Context,
+                                                           const MbedtlsPkcs11Credentials_t * pMbedtlsPkcs11Credentials )
+
+{
+    MbedtlsPkcs11Status_t returnStatus = MBEDTLS_PKCS11_SUCCESS;
+    int32_t mbedtlsError = 0;
+    CK_RV result = CKR_OK;
+
+    /* Parse the server root CA certificate into the SSL context. */
+    mbedtlsError = mbedtls_x509_crt_parse_file( &( pMbedtlsPkcs11Context->rootCa ),
+                                                pMbedtlsPkcs11Credentials->pRootCaPath );
+
+    if( mbedtlsError != 0 )
+    {
+        LogError( ( "Failed to parse server root CA certificate: mbedTLSError= %s : %s.",
+                    mbedtlsHighLevelCodeOrDefault( mbedtlsError ),
+                    mbedtlsLowLevelCodeOrDefault( mbedtlsError ) ) );
+        returnStatus = MBEDTLS_PKCS11_INVALID_CREDENTIALS;
+    }
+    else
+    {
+        mbedtls_ssl_conf_ca_chain( &( pMbedtlsPkcs11Context->config ),
+                                   &( pMbedtlsPkcs11Context->rootCa ),
+                                   NULL );
+        /* Setup the client private key. */
+        result = initializeClientKeys( pMbedtlsPkcs11Context,
```

Comment:
Use of client private key can be made optional to support use cases of basic TLS as well as HTTP requests to AWS services that use SigV4

Reply:
In those cases you wouldn't be using the PKCS11 transport interface as a key in the HSM is not used.

Reply:
But what if a customer wants to use the same TLS stack for HTTP operations with S3? They wouldn't want to have 2 different TLS stacks for MQTT and HTTP operations separately

Reply:
They would need a separate transport interface anyways. For that we could also have a mbedtls without pkcs11 transport interface. I think that is unrelated to this PR though.